The test seems to be failing on CPU: 

Traceback (most recent call last):
  File "jax/experimental/array_serialization/serialization_test.py", in test_transfer_shard_to_host
    sharding = SingleDeviceSharding(jax.devices()[0], memory_kind="device")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: Could not find memory addressable by device cpu. Device cpu can address the following memory kinds: unpinned_host. Got memory kind: device

Can you just skip on CPU?

--- a/jax/experimental/array_serialization/serialization.py
+++ b/jax/experimental/array_serialization/serialization.py
@@ -29,6 +29,7 @@ from typing import Any, Optional
 
 import jax
 from jax._src import array
+from jax._src import config
 from jax._src import distributed
 from jax._src import sharding
 from jax._src import sharding_impls
@@ -184,6 +185,22 @@ class _LimitInFlightBytes:
       self._cv.notify_all()
 
 
+def transfer_shard_to_host(shard: array.Shard) -> np.ndarray:
+  data = shard.data
+  has_pinned_host = any(
+      m.kind == "pinned_host" for m in shard.device.addressable_memories())
+  if config.enable_memories.value and has_pinned_host:
+    # If available, transfer to pinned host memory
+    sharding = jax.sharding.SingleDeviceSharding(shard.device,
+        memory_kind="pinned_host")
+    data = jax.device_put(data, sharding)
+    # Ensure that jax.Array's internal numpy array can be zero-copied.
+    # Tensorstore implicitly converts the written data to a numpy array,
+    # and would otherwise silently copy host-to-host.
+    data = np.array(data, copy=False)
+  return data
+
+
 async def async_serialize(
     arr_inp,
     tensorstore_spec,
@@ -260,8 +277,9 @@ async def async_serialize(
 
   async def _write_array(shard):
     if shard.replica_id == replica_id:
+      data = transfer_shard_to_host(shard)
       write_future = t[shard.index].write(
-          shard.data,
+          data,
           # Avoid additional copy of input array into the TensorStore chunk
           # cache.  If `arr_inp` is a jax.Array, the result of converting
           # it to a NumPy array, as is done internally by TensorStore, is

--- a/jax/experimental/array_serialization/serialization.py
+++ b/jax/experimental/array_serialization/serialization.py
@@ -194,13 +194,14 @@ async def transfer_shard_to_host(shard: array.Shard) -> np.ndarray:
     sharding = jax.sharding.SingleDeviceSharding(shard.device,
         memory_kind="pinned_host")
     data = jax.device_put(data, sharding)
-    # Allow other transfers to be scheduled simultaneously
-    await asyncio.sleep(0)
-    # Ensure that jax.Array's internal numpy array can be zero-copied.
-    # Tensorstore implicitly converts the written data to a numpy array,
-    # and would otherwise silently copy host-to-host.
-    data = np.array(data, copy=False)
-  return data
+  else:
+    data.copy_to_host_async()
+  # Allow other transfers to be scheduled simultaneously
+  await asyncio.sleep(0)
+  # Ensure that jax.Array's internal numpy array can be zero-copied. Tensorstore
+  # implicitly converts the written data to a numpy array, and would otherwise
+  # silently copy host-to-host.
+  return np.array(data, copy=False)
 
 
 async def async_serialize(

--- a/jax/experimental/array_serialization/serialization.py
+++ b/jax/experimental/array_serialization/serialization.py
@@ -185,7 +185,7 @@ class _LimitInFlightBytes:
       self._cv.notify_all()
 
 
-def transfer_shard_to_host(shard: array.Shard) -> np.ndarray:
+async def transfer_shard_to_host(shard: array.Shard) -> np.ndarray:
   data = shard.data
   has_pinned_host = any(
       m.kind == "pinned_host" for m in shard.device.addressable_memories())
@@ -194,6 +194,8 @@ def transfer_shard_to_host(shard: array.Shard) -> np.ndarray:
     sharding = jax.sharding.SingleDeviceSharding(shard.device,
         memory_kind="pinned_host")
     data = jax.device_put(data, sharding)
+    # Allow other transfers to be scheduled simultaneously
+    await asyncio.sleep(0)
     # Ensure that jax.Array's internal numpy array can be zero-copied.
     # Tensorstore implicitly converts the written data to a numpy array,
     # and would otherwise silently copy host-to-host.
@@ -277,7 +279,7 @@ async def async_serialize(
 
   async def _write_array(shard):
     if shard.replica_id == replica_id:
-      data = transfer_shard_to_host(shard)
+      data = await transfer_shard_to_host(shard)
       write_future = t[shard.index].write(
           data,
           # Avoid additional copy of input array into the TensorStore chunk

--- a/jax/experimental/array_serialization/serialization_test.py
+++ b/jax/experimental/array_serialization/serialization_test.py
@@ -28,7 +28,7 @@ import jax.numpy as jnp
 from jax._src import test_util as jtu
 from jax._src import array
 from jax._src import xla_bridge as xb
-from jax.sharding import NamedSharding, GSPMDSharding
+from jax.sharding import NamedSharding, GSPMDSharding, SingleDeviceSharding
 from jax.sharding import PartitionSpec as P
 from jax.experimental.array_serialization import serialization
 from jax.experimental.layout import Layout, DeviceLocalLayout as DLL
@@ -609,6 +609,21 @@ class CheckpointTest(jtu.JaxTestCase):
     out = deserialized_arr.astype(jnp.int8)  # doesn't crash
     self.assertEqual(out.dtype, jnp.int8)
     self.assertArraysEqual(out + out, out * 2)
+
+
+@jtu.with_config(jax_enable_memories=True)
+class TransferShardTest(jtu.JaxTestCase):
+
+  @jtu.skip_on_devices('cpu')
+  def test_transfer_shard_to_host(self):
+    np_inp = np.arange(16).reshape((4, 4))
+    sharding = SingleDeviceSharding(jax.devices()[0], memory_kind="device")
+    arr = jax.device_put(np_inp, sharding)
+    shard = arr.addressable_shards[0]
+
+    np_out = serialization.transfer_shard_to_host(shard)
+
+    self.assertArraysEqual(np_out, np_inp)
 
 if __name__ == '__main__':
   absltest.main(testLoader=jtu.JaxTestLoader())

